src/Tableau.cpp

```diff
@@ -0,0 +1,395 @@
+
+void Tableau::import(std::istream& is) {
+    tableau.clear();
+
+    std::size_t              nQubits = 0;
+    std::string              token;
+    std::string              line;
+    std::vector<std::string> data{};
+    char                     delimiter = '|';
+    // Try to find out size by reading first line
+    if (std::getline(is, line)) {
+        if (line.find('|', 0) == std::string::npos)
+            delimiter = ',';
+        parse_line(line, delimiter, {'\"'}, {'\\'}, data);
+        nQubits = static_cast<std::size_t>(std::stoul(data.at(0)));
+    }
+
+    tableau.reserve(nQubits);
+
+    while (std::getline(is, line)) {
+        tableau.emplace_back();
+        tableau.back().reserve(2 * nQubits + 1);
+        parse_line(line, delimiter, {'\"'}, {'\\', '\r', '\n', '\t'}, data);
+        bool skipFirst = true;
+        for (const auto& datum: data) {
+            if (skipFirst) {
+                skipFirst = false;
+                continue;
+            }
+            if (datum.empty())
+                continue;
+            tableau.back().emplace_back(static_cast<EntryType>(std::stoul(datum)));
+        }
+    }
+}
+
+void Tableau::populateTableauFrom(unsigned long bv, int nQubits,
+                                  int column) {
+    for (int j = 0; j < nQubits; ++j) {
+        if ((bv & (1U << j)) != 0U) {
+            tableau[j][column] = 1;
+        }
+    }
+}
+
+void Tableau::generateTableau(Tableau& tableau, const qc::QuantumComputation& circuit, int begin, int end) {
+    initTableau(tableau, circuit.getNqubitsWithoutAncillae());
+    int current_g = 0;
+    for (const auto& gate: circuit) {
+        if (current_g >= begin && (current_g < end || end < 0)) {
+            if (gate->getType() == qc::OpType::Compound) {
+                auto compOp = dynamic_cast<qc::CompoundOperation*>(gate.get());
+                auto cit    = compOp->begin();
+                while (cit != compOp->end() && current_g >= begin &&
+                       (current_g < end || end < 0)) {
+                    tableau.applyGate((*cit));
+                    ++cit;
+                    ++current_g;
+                }
+            } else {
+                tableau.applyGate(gate);
+                ++current_g;
+            }
+        }
+    }
+}
+
+void Tableau::initTableau(Tableau& tableau, std::size_t nqubits) {
+    tableau.init(nqubits);
+}
+
+int Tableau::applyGate(const std::unique_ptr<qc::Operation>& gate) {
+    auto nqubits = getQubitCount();
+    switch (gate->getType()) {
+        case qc::OpType::H: // HADAMARD
+        {
+            if (gate->isControlled()) {
+                util::fatal("Expected single-qubit gate");
+            }
+            const auto a = gate->getTargets().at(0U);
+            for (auto i = 0U; i < nqubits; i++) {
+                tableau[i][2U * nqubits] ^= (tableau[i][a] & tableau[i][a + nqubits]);
+                std::swap(tableau[i][a], tableau[i][a + nqubits]);
+            }
+            return 1U;
```

Comment:
please extract these updated into separate functions. there is so much code duplication in this whole function.